libsql-sqlite3/src/wal.c

```diff
@@ -4250,150 +4151,252 @@ static void libsqlGetWalPathname(char *buf, const char *orig, int orig_len) {
+#endif
+    out->methods.xDb = (void (*)(wal_impl *, sqlite3 *))sqlite3WalDb;
 
-  while (1) {
-    if (m == NULL) {
-      prev->pNext = pWalMethods;
-      pWalMethods->pNext = NULL;
-      break;
-    }
-    if (strcmp(m->zName, pWalMethods->zName) == 0) {
-      prev->pNext = pWalMethods;
-      pWalMethods->pNext = m->pNext;
-      break;
-    }
-    prev = m;
-    m = m->pNext;
+    WALTRACE(("WAL%d: opened\n", pRet));
   }
-  sqlite3_mutex_leave(mutex);
-  return SQLITE_OK;
+
+  return rc;
 }
 
-int libsql_wal_methods_unregister(libsql_wal_methods *pWalMethods) {
-  if (strncmp(pWalMethods->zName, "default", 7) == 0) {
-    return SQLITE_MISUSE;
-  }
-  MUTEX_LOGIC(sqlite3_mutex *mutex;)
-  MUTEX_LOGIC( mutex = sqlite3MutexAlloc(SQLITE_MUTEX_STATIC_MAIN); )
-  sqlite3_mutex_enter(mutex);
+void sqlite3DestroyWalManager(wal_manager_impl *self) { }
 
-  libsql_wal_methods *prev = libsql_wal_methods_find("default");
-  libsql_wal_methods *m = prev->pNext;
+int make_ref_counted_wal_manager(libsql_wal_manager wal_manager, RefCountedWalManager **out) {
+    RefCountedWalManager *p = (RefCountedWalManager*)sqlite3MallocZero(sizeof(RefCountedWalManager));
+    if (!p) return SQLITE_NOMEM;
+    p->n = 1;
+    p->ref = wal_manager;
+    p->is_static = 0;
+    *out = p;
+    return SQLITE_OK;
+}
 
-  while (m != NULL) {
-    if (strcmp(m->zName, pWalMethods->zName) == 0) {
-      prev->pNext = m->pNext;
-      break;
+/*
+ * Decrease the ref count and call the wal_manager destructor when the count reaches 0.
+ * Must be called from withing a critical section.
+ */
+void destroy_wal_manager(RefCountedWalManager *p) {
+    if (p->is_static) return;
+    assert(p->n != 0);
+    p->n -= 1;
+    if (p->n == 0) {
+        (p->ref.xDestroy)(p->ref.pData);
+        sqlite3_free(p);
     }
-    prev = m;
-    m = m->pNext;
-  }
-  sqlite3_mutex_leave(mutex);
-  return SQLITE_OK;
 }
 
-libsql_wal_methods *libsql_wal_methods_next(libsql_wal_methods *w) {
-  return w->pNext;
+/*
+ * Increase the ref count and return a pointer to the wal.
+ * Must be called from withing a critical section.
+ * Return NULL if the passed pointer ref count is already 0.
+ */
+RefCountedWalManager* clone_wal_manager(RefCountedWalManager *p) {
+    assert(p->n != 0);
+    p->n += 1;
+    return p;
 }
 
-const char *libsql_wal_methods_name(libsql_wal_methods *w) {
-  return w->zName;
+#define SQLITE3_WAL_MANAGER { \
+    .pData = NULL, \
+    .xOpen = (int (*)(wal_manager_impl *, sqlite3_vfs *, sqlite3_file *, int, long long, const char*, libsql_wal *))sqlite3WalOpen, \
+    .xClose = (int (*)(wal_manager_impl *, wal_impl *, sqlite3 *, int, int, unsigned char *))sqlite3WalClose, \
+    .bUsesShm = 1, \
+    .xLogDestroy = (int (*)(wal_manager_impl *, sqlite3_vfs*, const char*))sqlite3LogDestroy, \
+    .xLogExists = (int (*)(wal_manager_impl *, sqlite3_vfs*, const char*, int *))sqlite3LogExists, \
+    .xDestroy =(void (*)(wal_manager_impl*))sqlite3DestroyWalManager, \
```

Comment:
Can we get rid of this abomination? :sweat_smile: 
E.g. only instantiate `sqlite3_wal_manager_rc`, and use its `.ref` field wherever we need `sqlite3_wal_manager`? Passing it also becomes easier if you apply my suggestion about accepting it as a pointer in `sqlite3_open_v3` instead of passing it by value and copying it via stack.

Reply:
no, because we still need to expose the `sqlite3_wal_manager` for  external implementations that wrap vanilla sqlite WAL (like us)

Reply:
I don't follow - we could still expose it anywhere we want, and it's still instantiated in memory, it would just be available via the `sqlite3_wal_manager_rc` symbol.

Reply:
but that's an internal, we shouldn't expose that

Reply:
We shouldn't expose that ref counter, that's way too leaky

Reply:
an altenative is to make it a static that is initialized when `sqlite_` is called, let me try that